### PR TITLE
Load plugins from CMDB

### PR DESCRIPTION
**Pull request by roleyfoley**
## Description
Initial implementation of https://github.com/hamlet-io/engine/issues/1447 
This allows for a plugins to be sourced and loaded based on CMDB configuration. The plugin source configuration is managed at the layer level and can be used to reference plugins based on their hosting source ( git and local initially ) 

Example: 

If I have an account which I know is an aws based account I could add plugin loading configuration on the account layer object to load in the official hamlet aws plugin. We also want to support diagram drawing so we include that as well 

```json
{
    "Account" : {
        "Id" : "acct01",
        "Seed" : "abc123",
        "AWSId" : "1234567890",
        "Region" : "ap-southeast-2",
        "Plugins" : {
            "aws" : {
                "Enabled" : true,
                "Name" : "aws",
                "Source" : "git",
                "Priority" : 10,
                "Required" : true,
                "git" : {
                    "Url" : "https://github.com/hamlet-io/engine-plugin-aws",
                    "Ref" : "master",
                    "Path" : "aws/"
                }
            },
            "diagrams" : {
                "Enabled" : true,
                "Name" : "diagrams",
                "Source" : "git",
                "Priority" : 100,
                "Required" : true,
                "git" : {
                    "Url" : "https://github.com/hamlet-io/engine-plugin-diagrams",
                    "Ref" : "main",
                    "Path" : "diagrams/"
                }
            },
            "awsdiagrams" : {
                "Enabled" : true,
                "Name" : "awsdiagrams",
                "Source" : "git",
                "Priority" : 150,
                "Required" : true,
                "git" : {
                    "Url" : "https://github.com/hamlet-io/engine-plugin-aws",
                    "Ref" : "master",
                    "Path" : "awsdiagrams/"
                }
            }
        }
    }
}
``` 

Contract example 
```json
  "Stages": [
    {
      "Id": "pluginXinstall",
      "ExecutionMode": "parallel",
      "Mandatory": true,
      "Steps": [
        {
          "Id": "aws",
          "Type": "install_plugin",
          "Priority": 100,
          "Mandatory": true,
          "Parameters": {
            "Enabled": true,
            "Required": true,
            "Priority": 10,
            "Name": "aws",
            "Source": "git",
            "git": {
              "Configured": true,
              "Enabled": true,
              "Url": "https://github.com/hamlet-io/engine-plugin-aws",
              "Ref": "master",
              "Path": "aws/"
            },
            "local": {
              "Configured": true,
              "Enabled": true
            }
          }
        },
        {
          "Id": "diagrams",
          "Type": "install_plugin",
          "Priority": 100,
          "Mandatory": true,
          "Parameters": {
            "Enabled": true,
            "Required": true,
            "Priority": 100,
            "Name": "diagrams",
            "Source": "git",
            "git": {
              "Configured": true,
              "Enabled": true,
              "Url": "https://github.com/hamlet-io/engine-plugin-diagrams",
              "Ref": "main",
              "Path": "diagrams/"
            },
            "local": {
              "Configured": true,
              "Enabled": true
            }
          }
        },
        {
          "Id": "awsdiagrams",
          "Type": "install_plugin",
          "Priority": 100,
          "Mandatory": true,
          "Parameters": {
            "Enabled": true,
            "Required": true,
            "Priority": 150,
            "Name": "awsdiagrams",
            "Source": "git",
            "git": {
              "Configured": true,
              "Enabled": true,
              "Url": "https://github.com/hamlet-io/engine-plugin-aws",
              "Ref": "master",
              "Path": "awsdiagrams/"
            },
            "local": {
              "Configured": true,
              "Enabled": true
            }
          }
        }
      ]
    }
  ],
```

Using the `loader` entrance you can generate a contract which collects all of the plugins configured on different layers and gives an executor the information required to load the plugins and set them up. 

To make versioning possible the engine also accepts a plugin state file which provides results based on what the executor could find when loading plugins. This file is also used by the engine to confirm that a given plugin is available and if it is it will be loaded. If the plugin is a required plugin you will receive an error that the plugin is not loaded and should try again

## Motivation and Context
This allows for the plguin definitions to be shared between users and to make template generation consistent and manageable as plugins grow 

## How Has This Been Tested?
Tested locally 

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves the structure or operation of the implementation)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Followup Actions
- [x] None

## Checklist:
- [ ] My change requires a change to the [documentation](https://github.com/hamlet-io/docs).
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [X] None of the above.

### CODE REVIEW

**Comment by RossMurr4y** [comment]
I think a rebase is required, @gumbaba-gs implemented `Provider` (singular) in https://github.com/hamlet-io/engine/commit/ab555e5151128715f454ceef2ba9788cbe73f0f4

**Comment by roleyfoley** [comment]
Converting to draft based on #1480

**Comment by RossMurr4y** [comment]
The attributes `git` and `local` alongside capitalised attributes represents a departure from conventions we've used on all composite objects.

So far, attribute keys that are lowercase have all been `identifiers` (my word, because we call them Id's in-engine which is overused), which in-engine are identified by the parent attribute containing the property `SubObjects: true` which is not the case here.

Could we instead for now combine the `git` and `local` sections into one, and log a ticket to implement the JSONSchema concept of `oneOf` into the composite object structure. Together with https://github.com/hamlet-io/engine/issues/1491 (AttributeSets) this will allow composite objects to select from either the `git` schema or the `local` schema, but not both.

File: providers/shared/layers/layer.ftl [hunk 1]
```diff
@@ -59,3 +59,70 @@
...
+        {
+            "Names" : "Enabled",
+            "Description" : "To enable loading the plugin",
+            "Type" : BOOLEAN_TYPE,
+            "Default" : true
+        },
+        {
+            "Names" : "Required",
+            "Type" : BOOLEAN_TYPE,
+            "Description" : "Ensure the provider loads at all times",
```
Comment (RossMurr4y) [inline providers/shared/layers/layer.ftl:73]:
```suggestion
            "Description" : "Ensure the Plugin loads at all times",
```

**Comment by roleyfoley** [comment]
> The attributes `git` and `local` alongside capitalised attributes represents a departure from conventions we've used on all composite objects.
> 
> So far, attribute keys that are lowercase have all been `identifiers` (my word, because we call them Id's in-engine which is overused), which in-engine are identified by the parent attribute containing the property `SubObjects: true` which is not the case here.
> 
> Could we instead for now combine the `git` and `local` sections into one, and log a ticket to implement the JSONSchema concept of [`oneOf`](https://json-schema.org/understanding-json-schema/reference/combining.html#oneof) into the composite object structure. Together with #1491 (AttributeSets) this will allow composite objects to select from either the `git` schema or the `local` schema, but not both.

What does combining them achieve at this point in time? Wouldn't we then need to modify the schema again? 
We can just TitleCase the entry if that would be easier? Having them split was intentional to show what was available for a given source, combining them would make this harder to unserstand

**Comment by roleyfoley** [comment]
Based on our discussion I've renamed the `git` configuration parameter to `Source:git` with the idea that this represents configuration specific to the git source value. 

The local provider configuration has been removed in favour of the existing environment variable approach

**Comment by RossMurr4y** [comment]
I've tested in tandem with hamlet-io/executor-bash#134 and encountered two issues:
* If you don't have the aws/azure directories within the engine at all, it falls over during the generation contract pass.

```bash
master└─ $ mv /home/hamlet/cmdb/hamlet/engine/plugins/aws /home/hamlet/cmdb/hamlet/engine/plugins/_aws
hamlet @ ~/cmdb/cmdb/gs/msw-cmdb/indocsrm/config/solutionsv2/integration/default
master└─ $ $GENERATION_DIR/setup.sh

...

(Info) Generating generationcontract file ...

Exception in thread "main" java.io.FileNotFoundException: /home/hamlet/cmdb/hamlet/engine/plugins/aws does not exist.
        at freemarker.cache.FileTemplateLoader$1.run(FileTemplateLoader.java:123)
        at freemarker.cache.FileTemplateLoader$1.run(FileTemplateLoader.java:119)
        at java.security.AccessController.doPrivileged(Native Method)
        at freemarker.cache.FileTemplateLoader.<init>(FileTemplateLoader.java:119)
        at freemarker.cache.FileTemplateLoader.<init>(FileTemplateLoader.java:98)
        at io.codeontap.freemarkerwrapper.RunFreeMarker.main(RunFreeMarker.java:190)
```

* With the directories as-is, running `setup.sh` eventually fails after generating the profile contract with no steps.
```bash
master└─ $ $GENERATION_DIR/setup.sh
(Info) Generating generationcontract file ...

(Info) No change in generationcontract detected ...

(Info) Generating documents from generation contract /tmp/cot_LvpkYb/create_template_63mWpd/create_template_IGgpd4/results/loader-generation-contract.json
(Info) Generating plugincontract file ...

(Info) No change in plugincontract detected ...

(Info) No differences detected.
fatal: no upstream configured for branch 'feat-cmdb-providers'
fatal: No such remote ''
(Info) Loading plugins from contract...
(Info) [*] id: - name:
```

#### loader-plugincontract.json
```json
{
  "Metadata": {
    "Id": {},
    "Prepared": "2020-12-14T00:45:57Z",
    "RunId": "q2le75w36x",
    "RequestReference": "unassigned",
    "ConfigurationReference": "unassigned",
    "Providers": {}
  },
  "Stages": []
}
```

#### plugin-state.json
```json
{
  "Plugins": {
    "_engine": {
      "type": "git",
      "ref": "git::3a611b13",
      "plugin_dir": "/home/hamlet/cmdb/hamlet/engine/core"
    }
  }
}
```

#### Account.json
```
{
    "Account" : {
        ...
        "Plugins" : {
            "aws" : {
                "Enabled" : true,
                "Name" : "aws",
                "Priority" : 10,
                "Required" : true,
                "Source" : "git",
                "Source:git" : {
                    "Url" : "https://github.com/hamlet-io/engine-plugin-aws",
                    "Ref" : "master",
                    "Path" : "aws/"
                }
            },
            "diagrams" : {
                "Enabled" : true,
                "Name" : "diagrams",
                "Priority" : 100,
                "Required" : true,
                "Source" : "git",
                "Source:git" : {
                    "Url" : "https://github.com/hamlet-io/engine-plugin-diagrams",
                    "Ref" : "main",
                    "Path" : "diagrams/"
                }
            },
            "awsdiagrams" : {
                "Enabled" : true,
                "Name" : "awsdiagrams",
                "Priority" : 150,
                "Required" : true,
                "Source" : "git",
                "Source:git" : {
                    "Url" : "https://github.com/hamlet-io/engine-plugin-aws",
                    "Ref" : "master",
                    "Path" : "awsdiagrams/"
                }
            }
         }
    },
   ...
}
```

**Comment by roleyfoley** [comment]
@RossMurr4y This should now be fixed, it looks like a merge conflict removed the plugins configuration from the account layer. I was testing from loading plugins in the segment instead so missed it

**Review by RossMurr4y — APPROVED** [review]
Testing successful :)

**Pull request merged**